File: ios/RavenStorage/StorageModule.swift
import Foundation
import MMKV
import React

@objc(StorageModule)
public class StorageModule: NSObject {
    private static var instance: StorageModule? = nil

    // Private initializer for Singleton with encryption key
    private override init() {
        super.init()
        // Store this instance as the shared instance for native access
      StorageModule.setReactNativeInstance(self)
    }


    @objc
    public static func getInstance() -> StorageModule {
        if(instance == nil){
            instance = StorageModule()
        }
        return instance!
    }

    // Helper method to store and reuse the React Native-created instance
    @objc public static func setReactNativeInstance(_ instance: StorageModule) {
        if self.instance == nil {
            self.instance = instance
        }
    }
  
    @objc
    public func setString(storeId: String, forKey key: String, value: String) {
      StorageManager.shared.setString(storeId: storeId, key: key, value: value)
    }

    @objc
    public func initialize(storeId: String, encryptionKey:String, multiprocess:Bool) {
      StorageManager.shared.initialize(storeId: storeId,  encryptionKey: encryptionKey, multiprocess: multiprocess)
    }

    @objc
    public func getString(storeId: String, forKey key: String, defaultValue: String) -> String {
      let value = StorageManager.shared.getString(storeId: storeId, key: key, defaultValue: defaultValue) ?? defaultValue
        return value
    }


    @objc
    public func setBoolean(storeId: String, forKey key: String, value: Bool) {
      StorageManager.shared.setBoolean(storeId: storeId, key: key, value: value)
    }

    @objc public func getBoolean(storeId:String, forKey key: String, defaultValue: Bool) -> Bool {

      let value = StorageManager.shared.getBoolean(storeId: storeId, key: key, defaultValue: defaultValue)
        return value
    }


    @objc
    public func setNumber(storeId:String, forKey key: String, value: String) {
      StorageManager.shared.setString(storeId: storeId, key: key, value: value)
    }

    @objc
    public func getNumber(storeId:String, forKey key: String, defaultValue: String) -> String {
      let value = StorageManager.shared.getString(storeId: storeId, key: key, defaultValue: defaultValue)
      return value ?? defaultValue
    }


    @objc
  public func getAllKeys(storeId:String) -> [String] {
        return StorageManager.shared.getAllKeys(storeId: storeId)
    }

    @objc
    public func containsKey(storeId:String, forKey key: String) -> Bool {
      return StorageManager.shared.containsKey(storeId: storeId, key: key)
    }


    @objc
    public func removeKey(storeId:String, forKey key: String) {
      StorageManager.shared.removeKey(storeId: storeId, key: key)
    }

    // Remove all keys
    @objc
    public func removeAll(storeId:String) {
      StorageManager.shared.removeAll(storeId: storeId)
    }

    // Required for React Native
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
}

File: android/src/main/java/com/ravenstorage/StorageModule.kt
package com.ravenstorage

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class StorageModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  companion object {
    const val NAME = "StorageModule"
    
    init {
      try {
        System.loadLibrary("raven_storage")
      } catch (ignored: Exception) {
      }
    }
  }

  private external fun installStorageModule(runtimePtr: Long)

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun installJSIModule(): Boolean {
    val context = reactApplicationContext ?: return false
    val reactContext = context.javaScriptContextHolder
    val reactContextValue = reactContext?.get() ?: 0L

    return if (reactContextValue != 0L) {
      installStorageModule(reactContextValue)
      true
    } else {
      false
    }
  }

  fun initialize(storeId: String, encryptionKey: String, multiprocess: java.lang.Boolean) {
    StorageManager.initialize(
      reactApplicationContext,
      storeId,
      encryptionKey, multiprocess.booleanValue()
    )
  }

  fun setString(storeId: String, key: String, value: String) =
    StorageManager.setString(storeId, key, value)

  fun getString(
    storeId: String,
    key: String,
    defaultValue: String?
  ): String? = StorageManager.getString(storeId, key, defaultValue)

  fun setNumber(
    storeId: String,
    key: String,
    stringValue: String
  ) = StorageManager.setString(storeId, key, stringValue)

  fun getNumber(
    storeId: String,
    key: String,
    defaultValue: String
  ): String = StorageManager.getString(storeId, key, defaultValue) ?: defaultValue

  fun setBoolean(storeId: String, key: String, value: java.lang.Boolean) =
    StorageManager.setBoolRN(storeId, key, value)


  fun getBoolean(
    storeId: String,
    key: String,
    defaultValue: java.lang.Boolean?
  ): java.lang.Boolean? = StorageManager.getBoolRN(storeId, key, defaultValue)

  fun getAllKeys(storeId: String): ArrayList<String> =
    StorageManager.getAllKeys(storeId)

  fun removeAll(storeId: String) =
    StorageManager.removeAll(storeId)


  fun removeKey(storeId: String, key: String) =
    StorageManager.removeKey(storeId, key)

  fun containsKey(storeId: String, key: String): java.lang.Boolean =
    StorageManager.containsKey(storeId, key)


  override fun getName(): String {
    return "StorageModule"
  }
}

